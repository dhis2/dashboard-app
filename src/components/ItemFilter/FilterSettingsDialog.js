import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { DIMENSION_ID_PERIOD, DIMENSION_ID_ORGUNIT } from '@dhis2/analytics'
import i18n from '@dhis2/d2-i18n'
import {
    Button,
    ButtonStrip,
    Modal,
    ModalActions,
    ModalContent,
    ModalTitle,
    Radio,
    Transfer,
} from '@dhis2/ui'

import classes from './styles/FilterSettingsDialog.module.css'

const RadioOptions = ({
    filtersSelectable,
    updateFilterDimensionRestrictability,
}) => (
    <form>
        <div className={classes.radioButton}>
            <Radio
                dense
                checked={!filtersSelectable}
                label={i18n.t('Allow filtering by all dimensions')}
                name="radio-allow-filtering-on-all"
                onChange={payload => {
                    return updateFilterDimensionRestrictability(
                        payload.value === 'true'
                    )
                }}
                value={'false'}
            />
        </div>
        <div className={classes.radioButton}>
            <Radio
                dense
                checked={filtersSelectable}
                label={i18n.t('Only allow filtering by selected dimensions')}
                name="radio-restrict-filtering"
                onChange={payload => {
                    return updateFilterDimensionRestrictability(
                        payload.value === 'true'
                    )
                }}
                value={'true'}
            />
        </div>
    </form>
)

RadioOptions.propTypes = {
    filtersSelectable: PropTypes.bool,
    updateFilterDimensionRestrictability: PropTypes.func,
}

const FilterSettingsDialog = ({
    dimensions,
    restrictFilters,
    initiallySelectedItems,
    onClose,
    onConfirm,
    open,
}) => {
    const [selected, setSelected] = useState(initiallySelectedItems)
    const [filtersSelectable, setFiltersSelectable] = useState(restrictFilters)

    const updateFilterDimensionRestrictability = val => {
        if (val) {
            const otherItems = selected.filter(
                i => i !== DIMENSION_ID_ORGUNIT && i !== DIMENSION_ID_PERIOD
            )
            updateSelectedFilters([
                DIMENSION_ID_PERIOD,
                DIMENSION_ID_ORGUNIT,
                ...otherItems,
            ])
        }
        setFiltersSelectable(val)
    }

    const updateSelectedFilters = sItems => {
        // ensure that orgUnit, period appear before other items
        const periodItem = [
            ...new Set(sItems.filter(i => i === DIMENSION_ID_PERIOD)),
        ]
        const orgUnitItem = [
            ...new Set(sItems.filter(i => i === DIMENSION_ID_ORGUNIT)),
        ]
        const otherItems = sItems.filter(
            i => i !== DIMENSION_ID_ORGUNIT && i !== DIMENSION_ID_PERIOD
        )
        return setSelected([...periodItem, ...orgUnitItem, ...otherItems])
    }

    const closeDialog = () => {
        setSelected(initiallySelectedItems)
        setFiltersSelectable(restrictFilters)
        onClose()
    }

    return (
        <>
            {open && (
                <Modal large position="top" onClose={closeDialog}>
                    <ModalTitle>
                        {i18n.t('Dashboard filter settings')}
                    </ModalTitle>
                    <ModalContent>
                        <div className={classes.descContainer}>
                            <span>
                                {i18n.t(`Dashboards can be filtered by dimensions to change
                                         the data shown. By default, all dimensions are available
                                         as filters. Alternatively, only selected dimensions can
                                         be made available on a dashboard.`)}
                            </span>
                        </div>
                        <RadioOptions
                            filtersSelectable={filtersSelectable}
                            updateFilterDimensionRestrictability={
                                updateFilterDimensionRestrictability
                            }
                        />
                        {filtersSelectable && (
                            <Transfer
                                filterable
                                height="400px"
                                leftHeader={
                                    <div className={classes.transferHeader}>
                                        <span>
                                            {i18n.t('Available Filters')}
                                        </span>
                                    </div>
                                }
                                maxSelections={Infinity}
                                onChange={payload => {
                                    return updateSelectedFilters(
                                        payload.selected
                                    )
                                }}
                                options={dimensions.map(d => {
                                    return { label: d.name, value: d.id }
                                })}
                                optionsWidth="350px"
                                rightHeader={
                                    <div className={classes.transferHeader}>
                                        <span>
                                            {i18n.t('Selected Filters')}
                                        </span>
                                    </div>
                                }
                                selected={selected}
                                selectedWidth="350px"
                            />
                        )}
                    </ModalContent>
                    <ModalActions>
                        <ButtonStrip end>
                            <Button
                                onClick={closeDialog}
                                secondary
                                type="button"
                            >
                                {i18n.t('Cancel')}
                            </Button>
                            <Button
                                onClick={() => {
                                    if (!filtersSelectable) {
                                        setSelected([])
                                    }
                                    onConfirm(filtersSelectable, selected)
                                }}
                                primary
                                type="button"
                            >
                                {i18n.t('Confirm')}
                            </Button>
                        </ButtonStrip>
                    </ModalActions>
                </Modal>
            )}
        </>
    )
}

FilterSettingsDialog.propTypes = {
    dimensions: PropTypes.array,
    initiallySelectedItems: PropTypes.array,
    open: PropTypes.bool,
    restrictFilters: PropTypes.bool,
    onClose: PropTypes.func,
    onConfirm: PropTypes.func,
}

export default FilterSettingsDialog
