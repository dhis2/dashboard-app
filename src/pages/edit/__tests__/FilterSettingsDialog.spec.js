import { render } from '@testing-library/react'
import React from 'react'
import FilterSettingsDialog from '../FilterSettingsDialog'

jest.mock('@dhis2/app-runtime', () => ({
    useDhis2ConnectionStatus: () => ({ isConnected: true }),
}))

jest.mock('@dhis2/ui', () => {
    const originalModule = jest.requireActual('@dhis2/ui')

    return {
        __esModule: true,
        ...originalModule,
        Modal: function MockComponent(props) {
            return <div className="modal">{props.children}</div> //eslint-disable-line react/prop-types
        },
        Transfer: function MockTransfer(props) {
            return (
                <div
                    className="transfer"
                    /* eslint-disable react/prop-types, react/no-unknown-property */
                    options={props.options
                        .map(option => `${option.label}`)
                        .join(', ')}
                    selecteddimensions={props.selected.join(', ')}
                    /* eslint-enable react/prop-types, react/no-unknown-property */
                />
            )
        },
    }
})

jest.mock('../../../modules/useDimensions', () => ({
    __esModule: true,
    default: () => {
        return [
            { id: 'kl', name: 'kvikk lunsj' },
            { id: 'ss', name: 'salt sild' },
            { id: 'sm', name: 'seigmenn' },
        ]
    },
}))

test('renders correctly when filters are not restricted', () => {
    const { container } = render(
        <FilterSettingsDialog
            restrictFilters={false}
            initiallySelectedItems={[]}
            open={true}
        />
    )
    expect(container).toMatchSnapshot()
})

test('renders correctly when filters are restricted', () => {
    const { container } = render(
        <FilterSettingsDialog
            restrictFilters={true}
            initiallySelectedItems={['kl']}
            open={true}
        />
    )
    expect(container).toMatchSnapshot()
})
