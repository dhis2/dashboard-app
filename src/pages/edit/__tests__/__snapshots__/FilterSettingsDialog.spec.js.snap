// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when filters are not restricted 1`] = `
<div>
  <div
    class="modal"
  >
    <h1
      class="jsx-3701517325 title"
      data-test="dhis2-uicore-modaltitle"
    >
      Dashboard filter settings
    </h1>
    <div
      class="jsx-659511263 "
      data-test="dhis2-uicore-modalcontent"
    >
      <div
        class="descContainer"
      >
        <span>
          Dashboards can be filtered by dimensions to change
                                         the data shown. By default, all dimensions are available
                                         as filters. Alternatively, only selected dimensions can
                                         be made available on a dashboard.
        </span>
      </div>
      <form>
        <div
          class="radioButton"
        >
          <label
            class="jsx-1176042981 dense"
            data-test="dhis2-uicore-radio"
          >
            <input
              checked=""
              class="jsx-1176042981"
              name="radio-allow-filtering-on-all"
              type="radio"
              value="false"
            />
            <div
              class="jsx-1176042981 icon dense"
            >
              <svg
                class="jsx-1659897952 jsx-1631270792 checked"
                viewBox="0 0 14 14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  class="jsx-1659897952 jsx-1631270792 background"
                  cx="7"
                  cy="7"
                  r="7"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer unchecked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,13 C3.6862915,13 1,10.3137085 1,7 C1,3.6862915 3.6862915,1 7,1 C10.3137085,1 13,3.6862915 13,7 C13,10.3137085 10.3137085,13 7,13 Z"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer checked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,12 C4.23857625,12 2,9.76142375 2,7 C2,4.23857625 4.23857625,2 7,2 C9.76142375,2 12,4.23857625 12,7 C12,9.76142375 9.76142375,12 7,12 Z"
                />
                <circle
                  class="jsx-1659897952 jsx-1631270792 inner"
                  cx="7"
                  cy="7"
                  r="3"
                />
              </svg>
            </div>
            Allow filtering by all dimensions
          </label>
        </div>
        <div
          class="radioButton"
        >
          <label
            class="jsx-1176042981 dense"
            data-test="dhis2-uicore-radio"
          >
            <input
              class="jsx-1176042981"
              name="radio-restrict-filtering"
              type="radio"
              value="true"
            />
            <div
              class="jsx-1176042981 icon dense"
            >
              <svg
                class="jsx-1659897952 jsx-1631270792 "
                viewBox="0 0 14 14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  class="jsx-1659897952 jsx-1631270792 background"
                  cx="7"
                  cy="7"
                  r="7"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer unchecked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,13 C3.6862915,13 1,10.3137085 1,7 C1,3.6862915 3.6862915,1 7,1 C10.3137085,1 13,3.6862915 13,7 C13,10.3137085 10.3137085,13 7,13 Z"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer checked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,12 C4.23857625,12 2,9.76142375 2,7 C2,4.23857625 4.23857625,2 7,2 C9.76142375,2 12,4.23857625 12,7 C12,9.76142375 9.76142375,12 7,12 Z"
                />
                <circle
                  class="jsx-1659897952 jsx-1631270792 inner"
                  cx="7"
                  cy="7"
                  r="3"
                />
              </svg>
            </div>
            Only allow filtering by selected dimensions
          </label>
        </div>
      </form>
    </div>
    <div
      class="jsx-752323619"
      data-test="dhis2-uicore-modalactions"
    >
      <div
        class="jsx-71743532 end"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 secondary"
            data-test="dhis2-uicore-button"
            type="button"
          >
            Cancel
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="dhis2-uicore-button"
            type="button"
          >
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly when filters are restricted 1`] = `
<div>
  <div
    class="modal"
  >
    <h1
      class="jsx-3701517325 title"
      data-test="dhis2-uicore-modaltitle"
    >
      Dashboard filter settings
    </h1>
    <div
      class="jsx-659511263 "
      data-test="dhis2-uicore-modalcontent"
    >
      <div
        class="descContainer"
      >
        <span>
          Dashboards can be filtered by dimensions to change
                                         the data shown. By default, all dimensions are available
                                         as filters. Alternatively, only selected dimensions can
                                         be made available on a dashboard.
        </span>
      </div>
      <form>
        <div
          class="radioButton"
        >
          <label
            class="jsx-1176042981 dense"
            data-test="dhis2-uicore-radio"
          >
            <input
              class="jsx-1176042981"
              name="radio-allow-filtering-on-all"
              type="radio"
              value="false"
            />
            <div
              class="jsx-1176042981 icon dense"
            >
              <svg
                class="jsx-1659897952 jsx-1631270792 "
                viewBox="0 0 14 14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  class="jsx-1659897952 jsx-1631270792 background"
                  cx="7"
                  cy="7"
                  r="7"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer unchecked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,13 C3.6862915,13 1,10.3137085 1,7 C1,3.6862915 3.6862915,1 7,1 C10.3137085,1 13,3.6862915 13,7 C13,10.3137085 10.3137085,13 7,13 Z"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer checked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,12 C4.23857625,12 2,9.76142375 2,7 C2,4.23857625 4.23857625,2 7,2 C9.76142375,2 12,4.23857625 12,7 C12,9.76142375 9.76142375,12 7,12 Z"
                />
                <circle
                  class="jsx-1659897952 jsx-1631270792 inner"
                  cx="7"
                  cy="7"
                  r="3"
                />
              </svg>
            </div>
            Allow filtering by all dimensions
          </label>
        </div>
        <div
          class="radioButton"
        >
          <label
            class="jsx-1176042981 dense"
            data-test="dhis2-uicore-radio"
          >
            <input
              checked=""
              class="jsx-1176042981"
              name="radio-restrict-filtering"
              type="radio"
              value="true"
            />
            <div
              class="jsx-1176042981 icon dense"
            >
              <svg
                class="jsx-1659897952 jsx-1631270792 checked"
                viewBox="0 0 14 14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  class="jsx-1659897952 jsx-1631270792 background"
                  cx="7"
                  cy="7"
                  r="7"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer unchecked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,13 C3.6862915,13 1,10.3137085 1,7 C1,3.6862915 3.6862915,1 7,1 C10.3137085,1 13,3.6862915 13,7 C13,10.3137085 10.3137085,13 7,13 Z"
                />
                <path
                  class="jsx-1659897952 jsx-1631270792 outer checked"
                  d="M7,14 C10.8659932,14 14,10.8659932 14,7 C14,3.13400675 10.8659932,0 7,0 C3.13400675,0 0,3.13400675 0,7 C0,10.8659932 3.13400675,14 7,14 Z M7,12 C4.23857625,12 2,9.76142375 2,7 C2,4.23857625 4.23857625,2 7,2 C9.76142375,2 12,4.23857625 12,7 C12,9.76142375 9.76142375,12 7,12 Z"
                />
                <circle
                  class="jsx-1659897952 jsx-1631270792 inner"
                  cx="7"
                  cy="7"
                  r="3"
                />
              </svg>
            </div>
            Only allow filtering by selected dimensions
          </label>
        </div>
      </form>
      <div
        class="transfer"
        options="kvikk lunsj, salt sild, seigmenn"
        selecteddimensions="kl"
      />
    </div>
    <div
      class="jsx-752323619"
      data-test="dhis2-uicore-modalactions"
    >
      <div
        class="jsx-71743532 end"
        data-test="dhis2-uicore-buttonstrip"
      >
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 secondary"
            data-test="dhis2-uicore-button"
            type="button"
          >
            Cancel
          </button>
        </div>
        <div
          class="jsx-71743532 box"
        >
          <button
            class="jsx-2371629422 primary"
            data-test="dhis2-uicore-button"
            type="button"
          >
            Confirm
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
